### PR TITLE
Remove blocking

### PR DESCRIPTION
**Pull request by samuel-massinon**
Our non blocking calls didn't offer much except for calling `remotecall` which users can do themselves. Removing the option allows us to simplify the code.

https://github.com/invenia/FTPClient.jl/pull/23 should be merged first.

### CODE REVIEW

**Comment by samuel-massinon** [comment]
I think this PR makes more sense going to a new develop branch, since I'm going to do a bunch of changes. https://github.com/invenia/FTPClient.jl/pull/23 could be the last PR for version `0.1`, and this branch will start `0.2`.

**Comment by samuel-massinon** [comment]
@omus Ready to review.

File: src/FTPC.jl [hunk 1]
```diff
@@ -268,65 +256,58 @@ Download file with persistent connection.
...
-            process_response(ctxt, resp)
+        @ce_curl curl_easy_perform
+        process_response(ctxt, resp)
 
-            if ~isempty(save_path)
-                close(wd.buffer)
-            end
+        if ~isempty(save_path)
+            close(wd.buffer)
+        end
```
Comment (omus) [inline src/FTPC.jl]: Can this be moved into a `finally`?

File: src/FTPC.jl [hunk 2]
```diff
@@ -373,48 +350,41 @@ Upload file with persistent connection.
...
 
-            p_rd = pointer_from_objref(rd)
-            p_resp = pointer_from_objref(resp)
+        p_rd = pointer_from_objref(rd)
+        p_resp = pointer_from_objref(resp)
 
-            command = "STOR " * file_name
-            @ce_curl curl_easy_setopt CURLOPT_URL ctxt.url*file_name
-            @ce_curl curl_easy_setopt CURLOPT_CUSTOMREQUEST command
+        command = "STOR " * file_name
```
Comment (omus) [inline src/FTPC.jl]: Prefer string interpolation

File: src/FTPC.jl [hunk 3]
```diff
@@ -502,24 +472,18 @@ Establish connection to FTP server.
...
-    if options.blocking
-        ctxt = false
-        try
-            ctxt = setup_easy_handle(options)
-
-            # ping the server
-            resp = ftp_command(ctxt, "LIST")
+    ctxt = nothing
+    try
+        ctxt = setup_easy_handle(options)
```
Comment (omus) [inline src/FTPC.jl]: Do the setup prior to the `try`: http://docs.julialang.org/en/release-0.4/manual/control-flow/#finally-clauses

**Comment by omus** [comment]
Need to add documentation and examples on how to do non-blocking.

File: src/FTPC.jl (hunk 2 above)
Comment (samuel-massinon) [inline src/FTPC.jl]: In another branch I got rid of that command entirely.

**Comment by samuel-massinon** [comment]
@omus Do you think it would be best to update README per PR or at the end of all my changes? I feel like every PR changes part of the README and it'll be hard to get everything.

File: src/FTPC.jl [hunk 4]
```diff
@@ -346,20 +328,16 @@ Upload file with non-persistent connection.
...
 function ftp_put(file_name::AbstractString, file::IO, options::RequestOptions=RequestOptions())
-    if options.blocking
-        ctxt = false
-        try
+    ctxt = false
+    try
 
-            ctxt = setup_easy_handle(options)
-            resp = ftp_put(ctxt, file_name, file)
+        ctxt = setup_easy_handle(options)
```
Comment (omus) [inline src/FTPC.jl]: You should move all the `ctxt =` prior to the `try`

File: src/FTPC.jl [hunk 5]
```diff
@@ -243,18 +235,14 @@ Download file with non-persistent connection.
...
-
-            return resp
-        finally
-            cleanup_easy_context(ctxt)
-        end
-    else
-        return remotecall(ftp_get, myid(), file_name, set_opt_blocking(options), save_path)
+    ctxt = false
+    try
+        ctxt = setup_easy_handle(options)
```
Comment (omus) [inline src/FTPC.jl]: Put `ctxt =` prior to the `try`

File: src/FTPC.jl [hunk 6]
```diff
@@ -268,65 +256,59 @@ Download file with persistent connection.
...
 
-            return resp
+        return resp
 
-        catch
-            cleanup_easy_context(ctxt)
-            rethrow()
+    catch
+        cleanup_easy_context(ctxt)
+        rethrow()
```
Comment (omus) [inline src/FTPC.jl]: Should cleanup happen here? Isn't this similar to closing a file you never opened?

File: src/FTPC.jl [hunk 7]
```diff
@@ -502,24 +473,15 @@ Establish connection to FTP server.
...
-        catch
-            cleanup_easy_context(ctxt)
-            rethrow()
-        end
-    else
-        return remotecall(ftp_connect, myid(), set_opt_blocking(options))
+        return ctxt, resp
+    catch
+        cleanup_easy_context(ctxt)
+        rethrow()
```
Comment (omus) [inline src/FTPC.jl:458]: Again

File: src/FTPObject.jl [hunk 8]
```diff
@@ -72,44 +63,27 @@ end
...
-        return remotecall(upload_internal, myid(), ftp.ctxt, local_file, remote_name)
-    end
-end
-
-function upload_internal(ctxt::ConnContext, local_file::IO, remote_name::AbstractString)
+function upload(ftp::FTP, local_file::IO, remote_name::AbstractString)
     resp = nothing
     try
-        resp = ftp_put(ctxt, remote_name, local_file)
+        resp = ftp_put(ftp.ctxt, remote_name, local_file)
```
Comment (omus) [inline src/FTPObject.jl]: just `return` here

File: src/FTPObject.jl [hunk 9]
```diff
@@ -45,20 +45,11 @@ end
...
-        return remotecall(download_internal, myid(), ftp.ctxt, file_name, save_path)
-    end
-end
-
-function download_internal(ctxt::ConnContext, file_name::AbstractString, save_path::AbstractString)
+function download(ftp::FTP, file_name::AbstractString, save_path::AbstractString="")
     resp = nothing
     try
-        resp = ftp_get(ctxt, file_name, save_path)
+        resp = ftp_get(ftp.ctxt, file_name, save_path)
```
Comment (omus) [inline src/FTPObject.jl:52]: just `return` resp.body here. Why does this return `resp.body` while upload return `resp`?

**Comment by omus** [comment]
Ideally the PR would be smaller so the changes to the README wouldn't be as substantial

File: src/FTPObject.jl (hunk 9 above)
Comment (samuel-massinon) [inline src/FTPObject.jl:52]: I think we assumed that the body is what the user would care about.

Do you think users would like more info about the response, or should be just limit to the downlaod body and return nothing for upload?

File: src/FTPC.jl (hunk 7 above)
Comment (samuel-massinon) [inline src/FTPC.jl:458]: Again for the cleanup? `ctxt`  was created at the start of this function.

**Comment by omus** [comment]
LGTM +1

File: src/FTPC.jl (hunk 7 above)
Comment (omus) [inline src/FTPC.jl:458]: My bad. Why is this returning the `ctxt`?

File: src/FTPObject.jl (hunk 9 above)
Comment (omus) [inline src/FTPObject.jl:52]: Use your judgement here but I think upload should probably return nothing

File: src/FTPC.jl (hunk 7 above)
Comment (samuel-massinon) [inline src/FTPC.jl:458]: `ctxt` is how we maintain a persistent connection. Users use `ftp_connect` to make the connection and get `ctxt`.

Comment (omus) [inline src/FTPC.jl:458]: My bad. Totally reasonable. All of these functions are starting to look the same to me...

**Pull request merged**